compiler/liveness.go

```diff
@@ -1,54 +1,264 @@
+// or if there is a path from n to a node that uses a that does not contain a
+// re-definition of a.
+//
+// - A variable a is live-out at node n if it is live-in at one
+// of the successors of n.
 package compiler
 
-// FIXME: The current RegAlloc is based on wasm stack info and we probably
-// want a real one (in addition to this) with liveness analysis.
-// Returns the total number of registers used.
-func (c *SSAFunctionCompiler) RegAlloc() int {
-	regID := TyValueID(1)
-	valueRelocs := make(map[TyValueID]TyValueID)
-	for _, values := range c.StackValueSets {
-		for _, v := range values {
-			valueRelocs[v] = regID
-		}
-		regID++
-	}
-	for i := range c.Code {
-		ins := &c.Code[i]
-
-		if ins.Target != 0 {
-			if reg, ok := valueRelocs[ins.Target]; ok {
-				ins.Target = reg
-			} else {
-				panic("Register not found for target")
-			}
+import (
+	"github.com/go-interpreter/wagon/wasm"
+	"github.com/golang-collections/collections/stack"
+)
+
+// Set relative complement: S' = A ∖ B
+func setdiff(a, b []TyValueID) []TyValueID {
+	exclusion := make(map[TyValueID]bool)
+
+	for _, item := range b {
+		exclusion[item] = true
+	}
+
+	out := make([]TyValueID, 0)
+
+	for _, item := range a {
+		if _, isExcluded := exclusion[item]; !isExcluded {
+			out = append(out, item)
+		}
+	}
+
+	return out
+}
+
+func isLocalGetAction(instr Instr) bool {
+	// TODO(sven): use op constants
+	return instr.Op == "get_local"
+}
+
+func isLocalSetAction(instr Instr) bool {
+	// TODO(sven): use op constants
+	return instr.Op == "set_local" || instr.Op == "tee_local"
+}
+
+// Set union: S' = A ∪ B
+func union2(a, b []TyValueID) []TyValueID {
+	m := make(map[TyValueID]bool)
+
+	for _, item := range a {
+		m[item] = true
+	}
+
+	for _, item := range b {
+		if _, ok := m[item]; !ok {
+			a = append(a, item)
 		}
+	}
+
+	return a
+}
 
-		for j, v := range ins.Values {
-			if v != 0 {
-				if reg, ok := valueRelocs[v]; ok {
-					ins.Values[j] = reg
-				} else {
-					panic("Register not found for value")
+func isLoopEdge(parentBlock BasicBlock, block BasicBlock) bool {
+	// FIXME(sven): implement this
+	return false
+}
+
+// Variable definitions in a block
+func (liveness *LivenessProcessor) phiDefs(block BasicBlock) []TyValueID {
+	return liveness.locals
+}
+
+// Variable usage in a block
+func phiUses(block BasicBlock) []TyValueID {
+	values := make([]TyValueID, 0)
+	uses := make(map[int64]bool)
+
+	for _, instr := range block.Code {
+		if isLocalGetAction(instr) {
+			for _, value := range instr.Immediates {
+				if _, ok := uses[value]; !ok {
+					uses[value] = true
+					values = append(values, TyValueID(value))
 				}
 			}
 		}
 	}
 
-	return int(regID)
+	return values
 }
 
-func (ins *Instr) BranchTargets() []int {
-	switch ins.Op {
-	case "jmp", "jmp_if", "jmp_table":
-		ret := make([]int, len(ins.Immediates))
-		for i, t := range ins.Immediates {
-			ret[i] = int(t)
+// Identifies a block
+type blockID = int
```

Comment:
Alias isn't needed here.

Reply:
You mean this should be a regular type definition or I should just use int instead?

Reply:
I mean this: `type blockID int`